src/consul/watch.clj

```diff
@@ -110,85 +111,158 @@
+                  _ (log "Leader? " new-leader? result)
+                  continue? (or (= leader session)
+                                (async/>! leader-ch [k new-leader?]))]
+              (if continue?
+                (recur (assoc state :leader new-leader?))
+                (do (log "Finished and releasing:" k)
+                    (consul/kv-put conn k nil {:release session})
+                    (consul/session-destroy conn session)
+                    (async/>! leader-ch [k false]))))))))))
+
+(defn kill-leader!
+  "Given a channel representing a leader election node, give it a swift death."
+  [ch]
+  (async/close! ch)
+  (when-let [{:keys [conn k session]} (get @local-sessions ch)]
+    (consul/kv-put conn k nil {:release session})
+    (swap! local-sessions dissoc ch)))
+
+(defn kill-leader-pool!
+  "Attempt to evict all leaders and clear lock. Returns (assumed) success.
+   This will fail if there are any non-zombies, as they will reclaim the lock.
+   This works by attempting to evict sessions, and seeing if the session reconnects, so will spam stop/starts
+   for any non-zombies.
+   May return false positives if active leader has high latency.
+   Failure is not always negative.. it may mean a real leader has taken over from a zombie."
+  ([conn leader-key] (kill-leader-pool! conn leader-key 3))
+  ([conn leader-key n] (kill-leader-pool! conn leader-key n 300))
+  ([conn leader-key n next-leader-pause]
+   (not
+     (async/<!!
+       (async/go-loop [tries-left n]
+         (if (zero? tries-left)
+           true
+           (when-let [id (:session (consul.core/kv-get conn leader-key))]
+             (consul/kv-put :local leader-key nil {:release id})
+             (async/<! (async/timeout next-leader-pause))
+             (recur (dec tries-left)))))))))
+
+(comment
+  (def leader-key "/common/stuff/leader")
+
+  (def log-agent (agent nil))
+
+  (defn serial-log [& msgs]
+    (send-off log-agent #(apply println %2) msgs))
+
+  (defn close-component [var]
+    (when-let [old-ch (:ch @var)]
+      (println "closing..")
+      (kill-leader! old-ch))
+    (when-let [status (:status @var)]
+      (swap! status assoc :started false)))
+
+  (defn init-component [var id]
+    (close-component var)
+    (let [ch (async/chan)
+          status (or (:status @var) (atom {}))]
+      (reset! status {:started false :id id :bumps 0})
+      (leader-watch :local {:ttl "30m"} leader-key ch {:log (partial serial-log (str "[" id "]"))})
+      (async/go-loop []
+        (let [result (async/<! ch)]
+          (serial-log "Component:" id result)
+          (swap! status assoc :started (boolean (second result)))
+          (swap! status update :bumps inc)
+          (when result (recur))))
+      (alter-var-root var (constantly {:id id :ch ch :status status}))))
+
+  ;; BUGS:
+  ;; 1. when creating new session, seems to always fail (even if no leader). will work after timeout though
```

Comment:
Realise this bug report description is also more alarming than it should be:

"when immediately creating a new session, after own session is destroy without releasing the leader key, this node only will fail it's first attempt to reacquire the leader key using that new session, but on it's second loop it will be succesful"